### ISSUE TITLE
An employee should not be shared with their direct supervisor or shared with the same supervisor multiple times

### ISSUE THREAD

**Issue by Travis-A-Clark** [bcgov/performance#905]
Currently, an employee can show up as both a direct report and a shared employee for a supervisor. The employee can also be shared with the same supervisor multiple times. This means that they can appear multiple times in areas like the Conversation participant dropdown list, the Shared With or Audience fields of goals. 

We need to develop a process to:

1) Stop employees from being shared with their direct supervisor
2) Stop sharing employees that become direct reports at a later date (for example, employee A is shared with me in February; employee A then officially moves to my team in PeopleSoft in March and therefore shows up as my direct report; employee A should no longer be my shared employee)
3) Stop creating duplicates of shared employees that already exist (for example, if employee B is already shared with me, and someone tries to share employee B with me again, there should be no change made).

This process should apply to all shares created by supervisors, HR admins, or Sys admins.

**Comment by Travis-A-Clark** [comment]
@jizhaogit I added an additional item to the issue to prevent employees from being shared multiple times with the same supervisor. Let me know if you have any questions. Thanks!

**Comment by jizhaogit** [comment]
1. Stop employees from being shared with their direct supervisor  - **done**
2. Stop sharing employees that become direct reports at a later date (for example, employee A is shared with me in February; employee A then officially moves to my team in PeopleSoft in March and therefore shows up as my direct report; employee A should no longer be my shared employee) - **a schedule job to clean up this case after ods data pulling?  or modify ods data pulling to stop those sharing relationship?** 
3. Stop creating duplicates of shared employees that already exist (for example, if employee B is already shared with me, and someone tries to share employee B with me again, there should be no change made).  - **done**

**Comment by jizhaogit** [comment]
@telusdcinco please help do the code review about #1 and #3 and we can talk about #2 in the meeting or in our conversation

**Comment by SSubashree26** [comment]
@jizhaogit Hi Ji, 

Observation 1:
Noticed that the audience name is being repeated if a direct employee is shared to the same supervisor (while creating a goal from the goal bank) 
Data : For user Amy Miller

![Screenshot 2023-02-22 at 6.38.26 PM.png](https://images.zenhubusercontent.com/63892df90d58261296a2b5c0/84a57a82-5665-4da1-8e35-e1ce8575cafe)

Observation 2:
Also noticed that if an employee is being shared to the same supervisor again and again, the "Employee Profile Sharing" section displays repeated the information too.![Screenshot 2023-02-22 at 6.42.21 PM.png](https://images.zenhubusercontent.com/63892df90d58261296a2b5c0/fa1cee78-7d14-4ef1-bd06-84128d6245a1)

**Comment by jizhaogit** [comment]
added the same validation to system admin sharing and supervisor team member sharing.

**Comment by steffipravasi** [comment]
1)![Screenshot 2023-03-01 at 11.59.46 AM.png](https://images.zenhubusercontent.com/62e1b008ce7d6652fe102828/926eb975-b199-4b56-8af9-3bb62e1a23aa) 

2)![Screenshot 2023-03-01 at 11.57.13 AM.png](https://images.zenhubusercontent.com/62e1b008ce7d6652fe102828/68d6a538-fc05-4ac3-9dd2-accb9d76bd4c)

**Comment by jizhaogit** [comment]
@steffipravasi  @Travis-A-Clark Could you please help proved the content of the message for issue #2? thanks

**Comment by Travis-A-Clark** [comment]
@steffipravasi @jizhaogit For the two scenarios let's use the warning message:

1) The employee has already been shared with that supervisor. They cannot be shared with the same supervisor more than once.

2) The employee already reports directly to that supervisor. Employees cannot be shared with their direct supervisor.

**Comment by jizhaogit** [comment]
- applied the new message in the supervisor sharing part.
- corrected the grammar error about sysadmin/hradmin sharing

**Comment by steffipravasi** [comment]
@jizhaogit The warning messages has been updated on the Team Members page. Can we please update those 2 messages in HR Admin and SysAdmin - Share Employees page.

**Comment by steffipravasi** [comment]
Ticket has been validated. 
Messages are now updated for HR admin and Sys admin
![Screenshot 2023-03-13 at 11.14.25 AM.png](https://images.zenhubusercontent.com/62e1b008ce7d6652fe102828/31562e52-2b8b-4a7d-946e-dadc26fa131a)
![Screenshot 2023-03-13 at 11.06.21 AM.png](https://images.zenhubusercontent.com/62e1b008ce7d6652fe102828/bb181c04-1872-417d-b069-7fa1bb9fe2c3)
Moving ticket to PO review

**Comment by Travis-A-Clark** [comment]
@steffipravasi I am getting the wrong error message when I try to share an employee (Cathy May) with Travis Clark. The error message should be "The employee has already been shared with that supervisor. They cannot be shared with the same supervisor more than once." Instead I am seeing the other error message about direct supervisors (see below). Can you verify that the right message is displayed in each scenario? Thanks!

![image.png](https://images.zenhubusercontent.com/5ffe1f472e351dd9b643025b/6ae6f6b4-4b20-41b9-9c63-64a0199bcfaf)

**Comment by jizhaogit** [comment]
@Travis-A-Clark @steffipravasi  it is a copy/paste error when i update the latest message for the "my team member" sharing process.  a fix is applied for fixing this error in the test.  thanks!

**Comment by steffipravasi** [comment]
Ticket validated. 
For direct supervisors: The employee already reports directly to that supervisor. Employees cannot be shared with their direct supervisor. 
For shared employees: The employee has already been shared with that supervisor. They cannot be shared with the same supervisor more than once.
Validated at: My Team members, HRadmin-share, Sys admin - share

**Comment by jizhaogit** [comment]
i deleted the share relation about Davey,Heather and 	Wu,Jessica Hsin-Jung and tried to reproduce what @Travis-A-Clark did but can't get the warning message. 

@steffipravasi would you please try from your end and we can talk if you can reproduce the issue? thanks.

**Comment by Travis-A-Clark** [comment]
Looks like the warning message is gone. Steffi to confirm.

**Comment by steffipravasi** [comment]
@jizhaogit I Only noticed the error when the Share button in the pop-up confirmation modal is clicked multiple times. Can we restrict the user(HrAdmin/SysAdmin) to be allowed to click the Share button only once?

**Comment by jizhaogit** [comment]
@steffipravasi the quick multi-time clicking issue should be handled. would you take it a try?  thanks!

**Comment by jizhaogit** [comment]
javascript updated in hr/sys admin share employee modules.

**Comment by steffipravasi** [comment]
https://bcgov.sharepoint.com/:v:/r/teams/02915/Shared%20Documents/General/Testing/905.mov?csf=1&web=1&e=hCkGUt

**Comment by Travis-A-Clark** [comment]
@steffipravasi in your video, I don't see the share and cancel buttons in the confirmation pop-up (see below). For me, I have to click "share" in that modal before the process takes effect. Do you not have any option to cancel?

![image.png](https://images.zenhubusercontent.com/5ffe1f472e351dd9b643025b/525a6d8e-0bcb-44a5-90ef-426a4eae74a1)

**Comment by Travis-A-Clark** [comment]
Ignore my last comment.

@jizhaogit Once someone has clicked "Share" in the confirmation modal, can we add a note that says: "Processing request. Please do not close this window."

**Comment by jizhaogit** [comment]
message added

**Comment by steffipravasi** [comment]
Ticket Validated. 
Message 'Processing request. Do not close the window' is displayed after clicking on the Share button in the popup dialog box.
![Screenshot 2023-04-04 at 1.48.02 PM.png](https://images.zenhubusercontent.com/62e1b008ce7d6652fe102828/3b7a3d2a-344a-4754-8832-d5e70908bf0c)

**Comment by Travis-A-Clark** [comment]
@steffipravasi for some reason, I am not seeing the errors on the Sys Admin > Sharing screen anymore when I create a duplicate share. I see the processing request pop-up and then it goes through as if everyone worked fine. No error message like the screenshots you posted on March 13. Can you confirm?

**Comment by steffipravasi** [comment]
@Travis-A-Clark I have confirmed that the errors are displayed for Sys Admin
![Screenshot 2023-04-13 at 9.11.12 PM.png](https://images.zenhubusercontent.com/62e1b008ce7d6652fe102828/8e52a78b-9804-43f1-a841-0275668b86c5)![Screenshot 2023-04-13 at 9.13.02 PM.png](https://images.zenhubusercontent.com/62e1b008ce7d6652fe102828/6a64e2df-3335-49bc-9b3f-d04ff118752a)
Working fine on the team members too.
Although due to the site giving 504 error, I couldn't verify the same for HR admin.

**Comment by steffipravasi** [comment]
Was able to verify the issue for HR admin as well
![Screenshot 2023-04-14 at 3.08.12 PM.png](https://images.zenhubusercontent.com/62e1b008ce7d6652fe102828/45124265-e8d5-40e6-8b48-eeb3d6b4f8ab)![Screenshot 2023-04-14 at 3.10.59 PM.png](https://images.zenhubusercontent.com/62e1b008ce7d6652fe102828/d8307c38-bb6f-420d-a4f6-6fc33e71a36b)
